### ISSUE TITLE
Kill Defer.setTimeout

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11218]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1702|
|Type|task|
|Created|2006-05-05 06:55:42Z|
|Branch|https://github.com/twisted/twisted/tree/settimeout-1702-4|

It's been deprecated for long enough to kill. People are starting to ask questions.

spiv, I'll review it if you do it :)

Attachments:

* [timeout_1702.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9fe/9fed3a6d2f3211708e67b81f44901933cfe61e34/654c3fd3b7ca56b7ff7141f8621e2764d3a86bca.diff) (3848 bytes) - added by therve on 2007-01-12 11:48:58Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1702 1702
type__task task
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_settimeout_1702_4 branches/settimeout-1702-4
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1146812142000000 1146812142000000
changetime__1277901333000000 1277901333000000
version__ 
owner__ 
cc__therve cc__exarkun cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

An easy one. There seems to be a few remaining call through twisted, though:

 * spread/ui/gtk2util.py
 * spread/ui/gtkutil.py
 * names/dns.py
 * mail/alias.py

I'll try to see if they can be removed easily.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

I've removed spread and mail calls, and I've created [#2414](https://github.com/twisted/twisted/issues/2414) for dns.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|spiv commented|
|-|-|

What's holding this up?  I've looked at the patch, and it looks good to me.  Assuming tests pass, I'm happy for this to get merged.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to spiv|
|-|-|

That looks ready. Please review settimeout-1702-2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

So, hmm.

A couple problems jump out at me.

  * there's no test coverage for twisted/spread/ui/ at all.  Going strictly by the review process document, that's sufficient reason to not merge this as-is.  Since it's UI code, though...
  * however, the code was also broken previously, and is broken in the same way after this change.  if the brokenness were not innately tied to the use of setTimeout, this might not be a good reason to not merge the branch.  But it is. ;P  When the PB call eventually does complete, either successfully or with an error (eg ConnectionLost), it will blow up with an AlreadyCalledError.  To avoid this, it is necessary to add a layer of indirection through a second Deferred - similar to the way ThreadedResolver works.

The latter point is basically the same problem as exists in mail.py, which the branch actually does address, albeit in a somewhat inelegant manner (catching the AlreadyCalledError) ;)

So I'm not sure if I'm okaying this branch or rejecting it, really.  The mail changes certainly look correct and beneficial, maybe they should go in separately from the rest of this.  Someone who worked on twisted/spread/ui/ should probably look at those changes and maybe try to write some tests somehow.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to spiv|
|-|-|

I understand you concerns, but there is basically no chance that I ever write test code for this ui stuff (which should probably be rewrite to be actually testable).

I can probably live with that for now. If you create a ticket for the mail issue I'd be happy to extract the change.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

The mail issue got created at some point by someone and the mail parts of this taken care of.  The branch should be merged forward and the twisted/mail/ parts of it dropped (they conflict badly with what's in trunk now and are redundant).

As for the rest...  The gtkutil change _looks_ right, though it is untested.  We should probably start thinking about dropping our gtk1 code anyway.  However, the gtk2util change isn't as good as the gtkutil change and testing reveals that it indeed has a problem, 30 seconds after the connection is established, it logs an `AlreadyCalledError` traceback.  If it did the same thing gtkutil did I think it would work.

**Comment by twisted-trac** [comment]
|<img alt="rotund's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|rotund commented|
|-|-|

(In [26606]) Branching to 'remove-Defer-setTimeout-1702'

**Comment by twisted-trac** [comment]
|<img alt="rotund's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|rotund commented|
|-|-|

Modifying original branch

**Comment by twisted-trac** [comment]
|<img alt="rotund's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|rotund commented|
|-|-|

(In [26607]) Branching to 'settimeout-1702-3'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

[#4150](https://github.com/twisted/twisted/issues/4150) was a duplicate of this

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [29455]) Branching to 'settimeout-1702-4'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @therve|
|-|-|

I recovered the code from the deleted settimeout-1702-2 branch.  There were a bunch of conflicts, fortunately all resolved by picking the trunk version.

The branch needs a news fragment.  Manual testing of manhole shows that the `AlreadyCalledError` traceback I last reported has been fixed (does someone remember doing that?) so I think this should be merged now, missing tests notwithstanding (and we all know that we shouldn't make a habit of that).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set status to `closed`|
|-|-|

(In [29496]) Merge settimeout-1702-4

Authors: therve, rotund
Reviewer: exarkun
Fixes: [#11218](https://github.com/twisted/twisted/issues/11218)

Remove the long deprecated Deferred.setTimeout method.

A new era of peace has started.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
